src/util/address.rs

```diff
@@ -379,6 +379,7 @@ impl Display for Address {
                 version: ver,
                 program: ref prog,
             } => {
+                assert!(ver.to_u8() <= 16);
```

Comment:
If this is reachable by the user by feeding the API wrong data, it would be great to add a custom failure message explaining what failed and why, and how to check for that condition before to avoid a panic.

Now, what doesn't sit well with me is panicking in `Display`. That seems way to late, in a read-only function. If this address was invalid it should panic / error during construction / modification: right at the moment invalid data was crate, and not after the user already run with it on the other side of their application code.

Reply:
Since you can instantiate the Address struct the only way to catch before Display is by introducing a SegwitVersion type which encapsulates the u5 without allowing it to be greater than 16, is this desirable?

Reply:
Does `u5` mean this value can't really ever be greater than 16? In that case an assert just to double-check here would be perfectly fine, since it should not be user-triggereable.

Anyway - I'm a little bit short on familiarly with the code here and will let other people figure it out. My point is just that user should get a failure as soon as possible - during creation.

Reply:
no, u5 comes from bech32 and can be up to 32, like 2^5, also you can look at the test, I am instantiating it with 17 which is not defined as a possible segwit version (max 16).
Since the Address is created directly through a struct, I don't see any place you can get the failure sooner other than introducing the SegwitVersion type

Reply:
You could just make fields private, require going through a constructor and check it there.

Reply:
yes, you can make fields private but I would personally prefer keeping the struct initialization feature